### ISSUE TITLE
Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by shtse8** [coinfoundry/miningcore#11]
I saw another crash so I report here first. But I am still using yesterday version. I will try the latest fix today later.

```shell
[2017-09-24 08:12:58.9188] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:13:04.9204] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:13:05.9188] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:13:08.9257] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:13:10.9213] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:13:25.9628] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:13:55.9126] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:14:18.9170] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-24 08:14:20.9141] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
Segmentation fault (core dumped)
```

**Comment by shtse8** [comment]
```shell
[2017-09-24 13:07:28.5046] [D] [dash] [Pool] [0HL83544OO38B] Updating VarDiff
[2017-09-24 13:07:28.5046] [E] [dash] Error handling request: mining.submit System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'HandleContext has already been disposed'.
   at NetUV.Core.Native.NativeHandle.Validate()
   at NetUV.Core.Handles.Tcp.GetPeerEndPoint()
   at MiningCore.Blockchain.Bitcoin.BitcoinJobManager`2.<SubmitShareAsync>d__28.MoveNext() in /root/miningcore/src/MiningCore/Blockchain/Bitcoin/BitcoinJobManager.cs:line 355
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MiningCore.Blockchain.Bitcoin.BitcoinPoolBase`2.<OnSubmitAsync>d__6.MoveNext() in /root/miningcore/src/MiningCore/Blockchain/Bitcoin/BitcoinPoolBase.cs:line 156
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MiningCore.Blockchain.Bitcoin.BitcoinPoolBase`2.<OnRequestAsync>d__10.MoveNext() in /root/miningcore/src/MiningCore/Blockchain/Bitcoin/BitcoinPoolBase.cs:line 281
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MiningCore.Stratum.StratumServer`1.<>c__DisplayClass11_0.<<OnClientConnected>b__2>d.MoveNext() in /root/miningcore/src/MiningCore/Stratum/StratumServer.cs:line 160    at NetUV.Core.Native.NativeHandle.Validate()
   at NetUV.Core.Handles.Tcp.GetPeerEndPoint()
   at MiningCore.Blockchain.Bitcoin.BitcoinJobManager`2.<SubmitShareAsync>d__28.MoveNext() in /root/miningcore/src/MiningCore/Blockchain/Bitcoin/BitcoinJobManager.cs:line 355
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MiningCore.Blockchain.Bitcoin.BitcoinPoolBase`2.<OnSubmitAsync>d__6.MoveNext() in /root/miningcore/src/MiningCore/Blockchain/Bitcoin/BitcoinPoolBase.cs:line 156
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MiningCore.Blockchain.Bitcoin.BitcoinPoolBase`2.<OnRequestAsync>d__10.MoveNext() in /root/miningcore/src/MiningCore/Blockchain/Bitcoin/BitcoinPoolBase.cs:line 281
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MiningCore.Stratum.StratumServer`1.<>c__DisplayClass11_0.<<OnClientConnected>b__2>d.MoveNext() in /root/miningcore/src/MiningCore/Stratum/StratumServer.cs:line 160
[2017-09-24 13:07:28.5164] [E] [dash] [Pool] [0HL83544OO38B] Connection error state: [0HL83544OO38B] Incoming message exceeds maximum length of 8192
[2017-09-24 13:07:28.7360] [D] [JsonRpcConnection] [0HL83544OO38E] Received JsonRpc-Request: {"method": "mining.submit", "params": ["XrQsEqFfabd5mGy7W9vjBnvk6K19oc97gB.miner2", "331", "00000000", "59c7adf7", "26e48c8c"], "id":14}
[2017-09-24 13:07:28.7364] [D] [dash] [Pool] [0HL83544OO38E] Received request mining.submit [14]
```

**Comment by oliverw** [comment]
Did the log above result in a crash? It appears to be just diagnostic log spam.

**Comment by shtse8** [comment]
Should be two cases. Segmentation fault is another problem result in a crash. I just checked the log and discovered that exception.

**Comment by shtse8** [comment]
Suddenly segmentation fault (core dumped) again, but no detailed logs before segmentation fault.

**Comment by shtse8** [comment]
```
[2017-09-27 08:27:56.2079] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-27 08:28:32.2103] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-27 08:29:42.2078] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-27 08:30:13.2075] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-27 08:30:47.2060] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-27 08:31:19.2096] [I] [dgb-scrypt] [Bitcoin Job Manager] New block detected
[2017-09-27 08:31:25.2194] [I] [dash] [Pool] [0HL85FC11LCS9] Booting zombie-worker (idle-timeout exceeded)
Segmentation fault (core dumped)
```

**Comment by oliverw** [comment]
Yeah. It's now crystal clear that it always happens when the pool forcefully disconnectes a worker. Either through the banning system or because the client is considered dead. Working on it.

**Comment by bochom** [comment]
segmentation fault just after block found

**Comment by oliverw** [comment]
@shtse8 @bochom 
Observing similar crashes in production as well and got hold of enough crash dumps to work on the issue.

**Comment by oliverw** [comment]
Upstream issue is being [tracked here](https://github.com/StormHub/NetUV/issues/60).

**Comment by oliverw** [comment]
Fixed by d237e62a0e3ec6aaf9390db900477bec8e5bbb90

**Comment by oliverw** [comment]
Finally fixed by 6e35a735f611043ae8862bfbeff566e75753a33c
